watertap/core/zero_order_costing.py

```diff
@@ -0,0 +1,333 @@
+    # Unit operation costing methods
+    def exponential_flow_form(blk):
+        # Get parameter dict from database
+        parameter_dict = \
+            blk.unit_model.config.database.get_unit_operation_parameters(
+                blk.unit_model._tech_type,
+                subtype=blk.unit_model.config.process_subtype)
+
+        blk.capital_cost = pyo.Var(
+            initialize=1,
+            units=getattr(pyo.units, parameter_dict["capital_cost"]["units"]),
+            bounds=(0, None),
+            doc="Capital cost of unit operation")
+
+        t0 = blk.flowsheet().time.first()
+
+        # Get reference state for capital calculation
+        basis = parameter_dict["capital_cost"]["basis"]
+
+        # Get costing parameter sub-block for this technology
+        try:
+            # Try to get parameter Block from costing package
+            pblock = getattr(blk.config.flowsheet_costing_block,
+                             blk.unit_model._tech_type)
+        except AttributeError:
+            # If not present, call emthod to create parameter Block
+            pblock = _add_tech_parameter_block(blk, parameter_dict)
+
+        A = pblock.capital_a_parameter
+        B = pblock.capital_b_parameter
+
+        # Get state block for flow bases
+        try:
+            sblock = blk.unit_model.properties_in[t0]
+        except AttributeError:
+            # Pass-through case
+            sblock = blk.unit_model.properties[t0]
+
+        # TODO: More bases
+        if basis == "flow_vol":
+            state = sblock.flow_vol
+            state_ref = pblock.reference_state
+            sizing_term = state/state_ref
+        elif basis == "flow_mass":
+            state = sum(sblock.flow_mass_comp[j]
+                        for j in sblock.component_list)
+            state_ref = pblock.reference_state
+            sizing_term = state/state_ref
+        else:
+            raise ValueError(
+                f"{blk.name} - unrecognized basis in parameter declaration: "
+                f"{basis}.")
+
+        # TODO: Include TPEC/TIC
+        # TODO: Chemical addition
+        # TODO: Reduction parameter, uncertainty parameter
+        blk.capital_cost_constraint = pyo.Constraint(
+            expr=blk.capital_cost ==
+            A*pyo.units.convert(sizing_term,
+                                to_units=pyo.units.dimensionless)**B)
+
+        # Register flows if present
+        if hasattr(blk.unit_model, "electricity"):
+            blk.config.flowsheet_costing_block.cost_flow(
+                blk.unit_model.electricity[t0], "electricity")
+
+    # -------------------------------------------------------------------------
+    # Map costing methods to unit model classes
+    unit_mapping = {ZeroOrderBase: exponential_flow_form}
+
+
+def _add_tech_parameter_block(blk, parameter_dict):
+    # Parameters for this technology haven't been added yet
+    pblock = pyo.Block()
```

Comment:
Is the idea that this would support different forms? Or is this exponential form the standard for WT3?

Reply:
This method is only for exponential forms (which is the majority of the WT3 models). There are a few however which currently use custom forms (but those also tend to be the ones regressed from data).